Macro fixes

samscott89 opened the pull request:
I came across some cases where the `lsp_request!` and `lsp_notification!` macros returned missing types. I also added a couple of tests to help type-check these, and verify the macro names matched the method names.

**Note** that in addition to fixing the macros, I changed the `WillSave` enum to `WillSaveTextDocument` to be more consistent with the others. This has the potential to break crates depending on that enum, so if it's considered too late to change this, I can just fix the macro instead.

Marwes commented on src/notification.rs:
```diff
@@ -233,3 +236,47 @@ impl Notification for PublishDiagnostics {
+            if false {
+                // don't want to actually implement/run this
+                let params: <lsp_notification!($name) as Notification>::Params = unsafe { mem::uninitialized() };
+                fake_call::<lsp_notification!($name)>(params);
```
I think there is no need to create an uninitialized value here. Calling `fake_call::<lsp_notification!($name)>()` without any parameters should be enough to validate everything.

Marwes commented on src/request.rs:
```diff
@@ -447,3 +444,59 @@ impl Request for Rename {
+            if false {
+                // don't want to actually implement/run this
+                let params: <lsp_request!($name) as Request>::Params = unsafe { mem::uninitialized()};
+                let _res: <lsp_request!($name) as Request>::Result = fake_call::<lsp_request!($name)>(params);
```
Ditto

samscott89 commented:
You're right, not sure why I was thinking the args/return types were actually necessary to get Rust to type check. This version should have the same effect

Marwes commented:
Looks like one of the added checks are failing

```
warning: unused import: `std::mem`
   --> src/notification.rs:243:9
    |
243 |     use std::mem;
    |         ^^^^^^^^
    |
    = note: #[warn(unused_imports)] on by default
warning: unused import: `std::mem`
   --> src/request.rs:450:9
    |
450 |     use std::mem;
    |         ^^^^^^^^
error[E0277]: the trait bound `DocumentSymbol: request::Request` is not satisfied
   --> src/request.rs:463:24
    |
463 |             assert_eq!(<lsp_request!($name) as Request>::METHOD, $name);
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `request::Request` is not implemented for `DocumentSymbol`
...
485 |         check_macro!("textDocument/documentSymbol");
    |         -------------------------------------------- in this macro invocation
error[E0277]: the trait bound `DocumentSymbol: request::Request` is not satisfied
   --> src/request.rs:465:13
    |
465 |             fake_call::<lsp_request!($name)>();
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `request::Request` is not implemented for `DocumentSymbol`
...
485 |         check_macro!("textDocument/documentSymbol");
    |         -------------------------------------------- in this macro invocation
```

samscott89 commented:
Perfect! That's actually catching the change made in #73 which didn't update the macro. So the test is working as intended :)
I've rebased the branch on master, and fixed the macro in the latest push.

Marwes commented:
Thanks! Released as 0.50
